NO-JIRA: scaffold UI and add bootstrap and yarn.

dirkjlouw opened the pull request:
Just a scaffold for the UI. running angular 6 with yarn and installed bootstrap 4 beta 2 for build --prod in my exp does not give issues.

jannie-louwrens commented:
> bootstrap 4 beta 2 for build --prod
Bootstrap 4 is already on version 4.1.1 with a lot of patches and bug fixes. I would suggest you give it a try instead of the older beta version.

dirkjlouw commented:
@jannie-louwrens,

Thanks Jannie, I did bump it up. I added Font awesome and now removed it again. Will need to investigate small icons. 

Went ahead and made additional commits on this branch.

francois-opencollab commented:
Testing, do not reply

francois-opencollab commented:
Testing, do not reply pt2

trojanc commented:
> I added Font awesome and now removed it again. Will need to investigate small icons.
You could try clarity icons. Jannie suggested it for the Unisa project and it works very well. We've also used it for the vsslite upgrade. It works really easy and I've found that it has the least sizing issues compared to font-awesome or material icons (the icons always seemed to need a bunch of css magic to position and size it properly)

https://vmware.github.io/clarity/icons/clarity-icons

dirkjlouw commented:
[](https://www.dropbox.com/s/jiq6enyzeracfhf/Screenshot%202018-06-29%2010.09.30.png?dl=0)
[](https://www.dropbox.com/s/bp6ycc8fygi6mkr/Screenshot%202018-06-29%2010.09.22.png?dl=0)
[](https://www.dropbox.com/s/t3w83rxcajymaa3/Screenshot%202018-06-29%2010.09.13.png?dl=0)

dirkjlouw commented:
> I really would like us to start thinking about best practices (for OpenCollab), coding standards and styles, naming conventions, etc.

I came to like the naming convention for "dumb" objects be suffixing them with "Info". Something we have adopted from the Kuali projects and also started using in Assemble.EDU. For example in this case it could be CreditCardInfo. I've learned through repeatedly using this convention that it helps me to quickly grasp the meaning of a class/code by just looking at its name, i.e. *Info = represents some real world object, *Service = service contract, *Impl = some implementation of a service contract, etc.

The second thing I've learned or rather also came to like is to mirror the Java models in the Javascript world by having a similar name and sometimes content a like. I've found it helps a lot for any full stack developer to quickly identify and understand the code in all layers, especially if the name looks familiar.


@jannie-louwrens , Do not have a problem with this. Good practice. Would you like me to
1. Rename the ui/src/app/shared/model/creditcard.ts to creditcardinfo?

dirkjlouw commented on ui/src/app/app.module.ts:
```diff
@@ -4,11 +4,15 @@ import { NgModule } from '@angular/core';
   declarations: [
     AppComponent,
-    LandingComponent
+    LandingComponent,
```
Please have a look at the document paymentDocuments/PayGate PayWeb3 v1.03.pdf

It seems there are static content and thought it best to have a component. out of my exp I like to keep the home page in your reference clean to just run nav bar and footer. Landing page is something that is just information ald changes when the <router-outlet></router-outlet> change

jannie-louwrens replied:
I see, it sounds like you really want an About page, right? ;-)

trojanc replied:
`AppComponent` is the root of all Angular applications. It is always active with the other components nested in it. E.g. here is where Angular loads the first component to bootstrap the application https://github.com/OpenCollabZA/payment-simulator/blob/327cb9fe87a5530bd015f6dd373df0d7a9c09753/ui/src/index.html#L12

jannie-louwrens commented:
> Rename the ui/src/app/shared/model/creditcard.ts to creditcardinfo?

@dirkjlouw you can if you want to. 
All I wanted to underline is that we desperately need to conform to some OpenCollab standards now that more and more projects are started.

trojanc commented:
I don't know if you want to follow Angular's guide too (because it gets applied when you create content using the CLI)
https://angular.io/guide/styleguide#symbols-and-file-names

eg. according to their style guide, you'd call it `credit-card-info.ts` or `credit-card.info.ts` (if you want to use `info` as a type (like service, directive, etc)

trojanc commented on ui/pom.xml:
```diff
@@ -11,5 +11,63 @@
-</project>
\ No newline at end of file
+    <ui.maven.plugin.version>1.6</ui.maven.plugin.version>
+    <node.version>v8.10.0</node.version>
```
These are already defined in the parent pom. You could either use the parent's or delete the parents' if it makes more sense to have it at this module's level

trojanc commented on ui/src/app/creditcard/creditcard.component.html:
```diff
@@ -0,0 +1,77 @@
+          <div class="form-group">
+            <input type="text" class="form-control" id="cardholder" required placeholder="enter your name"
+                   [(ngModel)]="creditCard.cardholder" name="cardholder" #cardHolder="ngModel">
+            <div [hidden]="cardHolder.valid" class="alert alert-danger">required</div>
```
You could use Bootstrap's `invalid-feedback` indicators which creates a more subtle warning message for form validation.
 http://getbootstrap.com/docs/4.1/components/forms/#supported-elements

dirkjlouw replied:
These would have been sorted from having a design session. I will change it.

dirkjlouw replied:
Please provide me an example of a invalid-feedback or change one of the inputs on this code so I can see how they work. Tried for 30min to try and get them to work with no luck. Thank you

trojanc replied:
Quick and easiest way is to create a decorator that will apply the styles bootstrap requires

```typescript
import {Directive, HostBinding, Self} from '@angular/core';
import { NgControl } from '@angular/forms';

@Directive({
	selector : '[formControlName],[ngModel],[formControl]',
})
export class BSControlStatusDirective {

	public constructor(@Self() private control: NgControl) {
	}

	@HostBinding('class.is-valid')
	get ngClassValid(): boolean {
		if (this.control.control == null) {
			return false;
		}
		return this.control.control.valid;
	}

	@HostBinding('class.is-invalid')
	get ngClassInvalid(): boolean {
		if (this.control.control == null) {
			return false;
		}
		return this.control.control.invalid;
	}
}
```
I use this in another project. It is adapted from this Stackoverflow example https://stackoverflow.com/questions/46810764/angular-4-bootstrap-4-consolidating-validation-classes (I cleaned up that example to match Angular's style guide https://angular.io/guide/styleguide#style-06-03)

Add that directive to the module, and it should automatically apply the styling for bootstrap.
You can then just add the `.invalid-feedback` or `.valid-feedback` items right after the input control like in the bootstrap examples http://getbootstrap.com/docs/4.1/components/forms/#how-it-works

This solves atleast
- Red or Green input fields depending on the validity
- Feedback messages if required

It does not solve the issue if you wanted a specific message for a specific validation error (e.g minlength, required, format, etc)

trojanc commented on ui/src/app/app.component.html:
```diff
@@ -0,0 +1,74 @@
+  <router-outlet></router-outlet>
+  <hr>
+  <div class="landingfooterOptions">
+    <p>Powered by Paygate</p>
```
I'm not sure what my thoughts are here... We probably don't want people to think that this tool is created by Paygate... But we do somehow want to indicate that we are simulating paygate...

dirkjlouw replied:
@trojanc , I mocked the screens from the document. I could have just taken screenshots but saw it was just as quick to code it quick and dirty as we said in the Alt room. Again, would have been resolved if there was a design session

trojanc commented on ui/src/app/app.component.html:
```diff
@@ -0,0 +1,74 @@
+          <label>Paygate Test</label>
+        </div>
+      </div>
+      <div class="w-100"></div>
```
Why a blank column? Wouldn't just starting a new `.row` solve this?

dirkjlouw replied:
Again, design session. This is something I took off bootstrap components to wrap col and it looked nice.

trojanc commented on ui/src/app/app.component.html:
```diff
@@ -0,0 +1,74 @@
+        </div>
+      </div>
+      <div class="w-100"></div>
+      <div class="col">
```
It seems like you are on purpose creating too many columns to force then to wrap?
Isn't this a better way to indicate your intention with the layout?
```
<div class="form-group row>
  <div class="col">
      LABEL
  </div>
   <div class="col">
     FIELD
  </div>
</div>
<div class="form-group row>
  <div class="col">
      LABEL
  </div>
   <div class="col">
     FIELD
  </div>
</div>
```

dirkjlouw replied:
Again design session..... Thought the idea was quick and dirty.... but hey, will spend time changing it to above and see if it works

trojanc commented on ui/src/app/app.component.html:
```diff
@@ -0,0 +1,74 @@
+  <div class="container">
+    <div class="row">
+      <div class="col">
+        <div class="form-group landingLabels">
```
You could use the built in bootstrap utilities to align text instead of creating a new class just to right alight. Bootstrap has `.text-right` http://getbootstrap.com/docs/4.1/utilities/text/#text-alignment

dirkjlouw replied:
Who to say we will not have more css in the future? this can be debated right or left In my opinion

trojanc commented on ui/src/app/creditcard/creditcard.component.css:
```diff
@@ -0,0 +1,6 @@
+}
+.creditcardPaymentOptions{
+  text-align: center;
+}
```
You could have used bootstrap alignment utilities if alignment is all you need. http://getbootstrap.com/docs/4.1/utilities/text/#text-alignment

dirkjlouw replied:
same with previous session. If there are set standards sure why not.

trojanc commented on ui/src/app/eft/eft.component.css:
```diff
@@ -0,0 +1,6 @@
+}
+.eftPaymentOptions{
+  text-align: center;
+}
```
You could have saved quite a bit of duplicate code by using the provided alignment utilities http://getbootstrap.com/docs/4.1/utilities/text/#text-alignment

dirkjlouw replied:
sure.

trojanc commented on ui/src/app/shared/model/credit-card.info.ts:
```diff
@@ -0,0 +1,10 @@
+import {st} from "@angular/core/src/render3";
```
Remove this ghost import...?

dirkjlouw replied:
correct, ghost...

trojanc commented on ui/src/app/creditcard/creditcard.component.ts:
```diff
@@ -0,0 +1,19 @@
+import { Component, OnInit } from '@angular/core';
+import {Router} from "@angular/router";
```
Unused import

dirkjlouw replied:
maybe now but later maybe not

trojanc commented on ui/src/app/landing/landing.component.html:
```diff
@@ -0,0 +1,16 @@
+        <button type="button" routerLink="/creditcard" routerLinkActive="active" class="btn btn-outline-secondary">Credit Card</button>
+      </div>
+    </div>
+    <div class="w-100"></div>
```
Full column to force a new row? Why not use a `.row` with columns as intended?
```html
<div class="form-group row">
  <div class="col">
      LABEL
  </div>
   <div class="col">
     FIELD
  </div>
</div>
<div class="form-group row">
  <div class="col">
      LABEL
  </div>
   <div class="col">
     FIELD
  </div>
</div>
```

dirkjlouw replied:
again design session

dirkjlouw commented:
@trojanc , Please have a look at the lastest commit. I have amended all your suggestions.
